Makefile

```diff
@@ -0,0 +1,60 @@
+all: test check-coverage lint  ## test, check coverage and lint
+	@if [ -e .git/rebase-merge ]; then git --no-pager log -1 --pretty='%h %s'; fi
+	@echo '$(COLOUR_GREEN)Success$(COLOUR_NORMAL)'
+
+clean::  ## Remove generated files
+
+.PHONY: all clean
+
+
+# -- Test ---------------------------------------------------------------------
+
+COVERFILE = coverage.out
+COVERAGE = 100
+
+test:  ## Run tests and generate a coverage file
+	go test -coverprofile=$(COVERFILE) ./...
+
+check-coverage: test  ## Check that test coverage meets the required level
+	@go tool cover -func=$(COVERFILE) | $(CHECK_COVERAGE) || $(FAIL_COVERAGE)
+
+cover: test  ## Show test coverage in your browser
+	go tool cover -html=$(COVERFILE)
+
+clean::
+	rm -f $(COVERFILE)
+
+CHECK_COVERAGE = awk -F '[ \t%]+' '/^total:/ && $$3 < $(COVERAGE) {exit 1}'
+FAIL_COVERAGE = { echo '$(COLOUR_RED)FAIL - Coverage below $(COVERAGE)%$(COLOUR_NORMAL)'; exit 1; }
+
+.PHONY: test check-coverage cover
+
+
+# -- Lint ---------------------------------------------------------------------
+
+GOLINT_VERSION = 1.24.0
+GOLINT_INSTALLED_VERSION = $(or $(word 4,$(shell golangci-lint --version 2>/dev/null)),0.0.0)
+GOLINT_MIN_VERSION = $(shell printf '%s\n' $(GOLINT_VERSION) $(GOLINT_INSTALLED_VERSION) | sort -V | head -n 1)
+GOLINT_DOCKER = docker run --rm -v $(PWD):/src -w /src golangci/golangci-lint:v$(GOLINT_VERSION)  golangci-lint
+
+lint: ## Lint source code
+ifeq ($(GOLINT_MIN_VERSION), $(GOLINT_VERSION))
+	golangci-lint run
+else
+	$(GOLINT_DOCKER) run
```

Comment:
yeah, maybe.... I'm not sure about this.

Docker runs so much slower for me than the local install it isn't really a viable solution to me (maybe that's to do with caching I don't know).

I am almost tempted to throw it out altogether, but it works on CI right away which is nice, even though it ads 45 secs to the build.

It adds so much noise to the makefile as well which I really dislike.

The last thing I don't like is that we refer to the locally installed version once with `GOLINT_INSTALLED_VERSION` and then with `lint-with-local`...

Would you accept me throwing this out altogether?

Reply:
What you are proposing adds an extra two lines.... I might be ok if I just use: lint-with-docker I guess and get rid of the variable.... I don't know.

Reply:
The slowness of the docker version doesnt matter - if you dont want it slow, just make sure you have the right version installed locally and it wont use docker. I am not suggesting only the docker version, but that it should always be available with `make lint-with-docker` or some target.

I need the docker version because i dont have it installed locally. For me it is not slow, because lint does not rebuild anything. It is a problem on Macs because the stuff you build is built for Darwin, but docker runs Linux, so everything is rebuilt. Pretty sure I don't see this slowness.

But another reason I always want a `lint-with-docker` target is that it is consistent for everyone. Relying on locally installed software that revs as fast as golangci-lint does, means everyone has a different version. When trying to diagnose problems that other's have, having a consistent version that everyone can run helps a lot.